// eslint-plugin-regexp/dist/rules/no-assertion-capturing-group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.default = utils_1.createRule("no-assertion-capturing-group", {
    meta: {
        docs: {
            description: "disallow capturing group that captures assertions.",
            recommended: true,
        },
        schema: [],
        messages: {
            unexpected: "Unexpected capture assertions.",
        },
        type: "suggestion",
    },
    create(context) {
        const sourceCode = context.getSourceCode();
        function createVisitor(node) {
            return {
                onCapturingGroupEnter(cgNode) {
                    if (cgNode.alternatives.every((alt) => alt.elements.every((e) => e.type === "Assertion"))) {
                        context.report({
                            node,
                            loc: utils_1.getRegexpLocation(sourceCode, node, cgNode),
                            messageId: "unexpected",
                        });
                    }
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
