// eslint-plugin-regexp/dist/rules/no-useless-exactly-quantifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.default = utils_1.createRule("no-useless-exactly-quantifier", {
    meta: {
        docs: {
            description: "disallow unnecessary exactly quantifier",
            recommended: true,
        },
        schema: [],
        messages: {
            unexpected: 'Unexpected quantifier "{{expr}}".',
        },
        type: "suggestion",
    },
    create(context) {
        const sourceCode = context.getSourceCode();
        function createVisitor(node) {
            return {
                onQuantifierEnter(qNode) {
                    if (qNode.min === qNode.max &&
                        (qNode.min === 0 || qNode.min === 1)) {
                        const [startOffset, endOffset] = utils_1.getQuantifierOffsets(qNode);
                        const text = qNode.raw.slice(startOffset, endOffset);
                        context.report({
                            node,
                            loc: utils_1.getRegexpLocation(sourceCode, node, qNode, [
                                startOffset,
                                endOffset,
                            ]),
                            messageId: "unexpected",
                            data: {
                                expr: text,
                            },
                        });
                    }
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
