// eslint-plugin-regexp/dist/rules/no-empty-group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.default = utils_1.createRule("no-empty-group", {
    meta: {
        docs: {
            description: "disallow empty group",
            recommended: true,
        },
        schema: [],
        messages: {
            unexpected: "Unexpected empty group.",
        },
        type: "suggestion",
    },
    create(context) {
        const sourceCode = context.getSourceCode();
        function verifyGroup(node, gNode) {
            if (gNode.alternatives.every((alt) => alt.elements.length === 0)) {
                context.report({
                    node,
                    loc: utils_1.getRegexpLocation(sourceCode, node, gNode),
                    messageId: "unexpected",
                });
            }
        }
        function createVisitor(node) {
            return {
                onGroupEnter(gNode) {
                    verifyGroup(node, gNode);
                },
                onCapturingGroupEnter(cgNode) {
                    verifyGroup(node, cgNode);
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
