// eslint-plugin-regexp/dist/rules/no-octal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.default = utils_1.createRule("no-octal", {
    meta: {
        docs: {
            description: "disallow octal escape sequence",
            recommended: true,
        },
        schema: [],
        messages: {
            unexpected: 'Unexpected octal escape sequence "{{expr}}".',
        },
        type: "suggestion",
    },
    create(context) {
        const sourceCode = context.getSourceCode();
        function createVisitor(node) {
            return {
                onCharacterEnter(cNode) {
                    if (cNode.raw.startsWith("\\0") && cNode.raw !== "\\0") {
                        context.report({
                            node,
                            loc: utils_1.getRegexpLocation(sourceCode, node, cNode),
                            messageId: "unexpected",
                            data: {
                                expr: cNode.raw,
                            },
                        });
                    }
                },
            };
        }
        return utils_1.defineRegexpVisitor(context, {
            createVisitor,
        });
    },
});
